src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -353,35 +240,226 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+            }
+          }
+        }
+        case (x: Cond, _) => Some(compileCond(x))
+        case (x: QuantifiedBody, index) => {
+          val newIndexPrefix = s"${indexPrefix}${index}_"
+          val subqueryWhereStr = generateWhereClause(x.bodies, newIndexPrefix)
+          val subqueryFromStr  = generateFromClause(x.bodies, newIndexPrefix)
+          x.modifier match {
+            case ExistModifier(negated) => Some(s"${if (negated) "NOT " else ""}EXISTS (SELECT 1 FROM ${subqueryFromStr} WHERE ${subqueryWhereStr})")
+            case AllModifier() => Some(s"NOT EXISTS (SELECT 1 FROM ${subqueryFromStr} WHERE NOT (${subqueryWhereStr}))")
+            case _ => None
+          }
+        }
+      }
+      conditions.mkString(" AND ")
+    }
+
+    def generateFromClause(bodies: List[Body], indexPrefix: String) : String = {
+      val atoms = bodies.zipWithIndex flatMap {
+        case (x:Atom,i) => Some(s"${x.name} R${indexPrefix}${i}")
+        case _ => None
+      }
+      val outers = bodies.zipWithIndex flatMap {
+        case (x:QuantifiedBody,i) => {
+          val newIndexPrefix = s"${indexPrefix}${i}_"
+          x.modifier match {
+            case OuterModifier() => {
+              val from = generateFromClause(x.bodies, newIndexPrefix)
+              val joinCond = generateWhereClause(x.bodies, newIndexPrefix)
+              Some(s"${from} ON ${joinCond}")
+            }
+            case _ => None
+          }
+        }
+        case _ => None
+      }
+      // full outer join
+      if (atoms isEmpty) {
+        outers.mkString(" FULL OUTER JOIN ")
+      } else {
+        (atoms.mkString(", ") +: outers).mkString(" LEFT OUTER JOIN ")
+      }
+    }
+
+    def optionalClause(prefix: String, clause: String): String =
+      if (clause.isEmpty) "" else prefix + " " + clause
+
+    // check if an expression contains an aggregation function
+    def containsAggregation(expr: Expr) : Boolean = {
+      expr match {
+        case VarExpr(name) => false
+        case ConstExpr(value) => false
+        case FuncExpr(function, args, agg) => if (agg) agg else {
+          args.map(containsAggregation).foldLeft(false)(_ || _)
+        }
+        case BinaryOpExpr(lhs, op, rhs) => containsAggregation(lhs) || containsAggregation(rhs)
+        case TypecastExpr(lhs, rhs) => containsAggregation(lhs)
+      }
+    }
+
+    // handle group by
+    // map head terms, leaving out aggregation functions
+    val groupbyTerms = cq.head.terms.filterNot(containsAggregation) map (compileExpr(_))
+
+    val groupbyStr = if (groupbyTerms.size == cq.head.terms.size || groupbyTerms.isEmpty) {
+      ""
+    } else {
+      s"\n        GROUP BY ${groupbyTerms.mkString(", ")}"
+    }
+
+    // limit clause
+    val limitStr = optionalClause("LIMIT", cq.limit.getOrElse("").toString)
+
+    s"""FROM ${ generateFromClause(cq.bodies(0), "") }
+        ${ optionalClause("WHERE", generateWhereClause(cq.bodies(0), "")) }${groupbyStr}${limitStr}"""
```

Comment:
It's still a mystery to me how this works with `bodies(0)`.  It may be a good idea to parameterize that index or redefine the function as `Body => String` instead.  Otherwise, I think we should compile a UNION of all bodies.